AdvTester: Second of two tests fails when run together---each works when run individually

I've tried to do some parameterized design in chisel3.
I've defined two tests which use testbenches and implemention of the same module but with different parameters. It seems that the implementation in the second test is not elaborated (caused the implementation with different/wrong parameters to be used.)

Here is a complete file that will expose this issue.

````
package bug

import chisel3._
import chisel3.iotesters._
import org.scalatest.GivenWhenThen
import scala.runtime.ScalaRunTime.stringOf

class CarryChainIfc(N:Int,width:Int) extends Module {
   val io = new Bundle {
      val a  = Vec( N, UInt(INPUT,width=width))
      val co = Vec( N, UInt(OUTPUT,width=width))
   }
}

class ORChain(N:Int) extends CarryChainIfc(N,1) {
   require( io.a.length == io.co.length)
   require( io.a.length == N)
   io.co(0) := io.a(0)
   for( ((a,ci),co)<-io.a.toList.tail.zip( io.co).zip( io.co.tail)) {
      co := a | ci
   }
}

abstract class Action
case class ActionInputArray(val nm:String, val s:Array[BigInt])  extends Action
case class ActionOutputArray(val nm:String, val s:Array[BigInt])  extends Action
case class ActionDelay(val v:Int)  extends Action

class CarryChainTests[T <: CarryChainIfc]( ActList:List[Action], c: T) extends AdvTester(c,verbose=true)
{

  for (a <- ActList) {  
    a match {
    case ActionInputArray("a",s) => {
       require( s.length == c.io.a.length)
       for ( (f,a) <- c.io.a.zip( s)) {
          reg_poke( f, a)
       }
    }
    case ActionOutputArray("co",s) => {
       require( s.length == c.io.co.length)
       for ( (f,a) <- c.io.co.zip( s)) {
          expect( f, a)
       }
    }
    case ActionDelay(v) => takesteps(1)()
    }
  }
}

class CarryChainTester[ +T <: CarryChainIfc]( val dut: () => T) extends ChiselFlatSpec with GivenWhenThen {
//  val backend = "firrtl"
  val backend = "verilator"
  val action_buffer= new collection.mutable.ListBuffer[Action]()

  def InputArray( nm:String, tag:String, s:Array[BigInt]):String = {
     action_buffer += new ActionInputArray(nm,s)
     s"InputArray $nm $tag ${stringOf(s)}"
  }

  def OutputArray( nm:String, tag:String, s:Array[BigInt]):String = {
     action_buffer += new ActionOutputArray(nm,s)
     s"OutputArray $nm $tag ${stringOf(s)}"
  }
  
  def Delay(v:Int):String = {
     action_buffer += new ActionDelay(v)
     s"After $v cycle" + (if ( v != 1) "s" else "")
  }
  def init():Unit = action_buffer.clear
  def run():Boolean = {
     chisel3.iotesters.Driver(dut,backend) {
       (component) =>
           new CarryChainTests( action_buffer.toList,component)
     }
  }
}

class ORChainTester(val N:Int) extends CarryChainTester( ()=>new ORChain(N)) {
  def v(bin: String): Array[BigInt] = {
     assert( N === bin.length)
     bin.toList.map( "01".indexOf(_)).map( BigInt(_)).reverse.toArray
  }
}


class ORChainTest2 extends ORChainTester(2) {
  s"ORChain of size ${N}" should s"work with various inputs" in {
    init()
    Given("a ORChain module")
    for ( (a,co) <- List( (v("00"),v("00")),
                          (v("01"),v("11")),
                          (v("10"),v("10")),
                          (v("11"),v("11")))) {
      When(InputArray("a","is",a))
      And(Delay(1))
      Then(OutputArray("co","is",co))
    }
    run() should be (true)
  }
}

class ORChainTest4 extends ORChainTester(4) {
  s"ORChain of size ${N}" should s"work with various inputs" in {
    init()
    Given("a ORChain module")
    for ( (a,co) <- List( (v("0000"),v("0000")),
                          (v("0001"),v("1111")),
                          (v("0010"),v("1110")),
                          (v("0100"),v("1100")),
                          (v("1000"),v("1000")))) {
      When(InputArray("a","is",a))
      And(Delay(1))
      Then(OutputArray("co","is",co))
    }
    run() should be (true)
  }
}
````

I get this LOG file with a compiler error when the second module is compiled (it then continues to run the wrong test.)
````
smburns@smburns-VirtualBox:~/bug-parameterized-dut$ sbt test
[info] Set current project to SCL PPT Testing Experiments (in build file:/home/smburns/bug-parameterized-dut/)
[info] [0.004] Elaborating design...
[info] [0.287] Done elaborating.
verilator --cc ORChain.v --assert -Wno-fatal -Wno-WIDTH -Wno-STMTDLY --trace -O2 --top-module ORChain +define+TOP_TYPE=VORChain +define+PRINTF_COND=!ORChain.reset +define+STOP_COND=!ORChain.reset -CFLAGS -Wno-undefined-bool-conversion -O2 -DTOP_TYPE=VORChain -include VORChain.h -Mdir test_run_dir/bug.ORChain --exe ORChain-harness.cpp
make: Entering directory `/home/smburns/bug-parameterized-dut/test_run_dir/bug.ORChain'
g++  -I.  -MMD -I/usr/local/share/verilator/include -I/usr/local/share/verilator/include/vltstd -DVL_PRINTF=printf -DVM_TRACE=1 -DVM_COVERAGE=0 -Wno-char-subscripts -Wno-parentheses-equality -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable     -Wno-undefined-bool-conversion -O2 -DTOP_TYPE=VORChain -include VORChain.h   -c -o ORChain-harness.o ORChain-harness.cpp
g++  -I.  -MMD -I/usr/local/share/verilator/include -I/usr/local/share/verilator/include/vltstd -DVL_PRINTF=printf -DVM_TRACE=1 -DVM_COVERAGE=0 -Wno-char-subscripts -Wno-parentheses-equality -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable     -Wno-undefined-bool-conversion -O2 -DTOP_TYPE=VORChain -include VORChain.h   -c -o verilated.o /usr/local/share/verilator/include/verilated.cpp
g++  -I.  -MMD -I/usr/local/share/verilator/include -I/usr/local/share/verilator/include/vltstd -DVL_PRINTF=printf -DVM_TRACE=1 -DVM_COVERAGE=0 -Wno-char-subscripts -Wno-parentheses-equality -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable     -Wno-undefined-bool-conversion -O2 -DTOP_TYPE=VORChain -include VORChain.h   -c -o verilated_vcd_c.o /usr/local/share/verilator/include/verilated_vcd_c.cpp
/usr/bin/perl /usr/local/share/verilator/bin/verilator_includer -DVL_INCLUDE_OPT=include VORChain.cpp > VORChain__ALLcls.cpp
/usr/bin/perl /usr/local/share/verilator/bin/verilator_includer -DVL_INCLUDE_OPT=include VORChain__Trace.cpp VORChain__Syms.cpp VORChain__Trace__Slow.cpp > VORChain__ALLsup.cpp
g++  -I.  -MMD -I/usr/local/share/verilator/include -I/usr/local/share/verilator/include/vltstd -DVL_PRINTF=printf -DVM_TRACE=1 -DVM_COVERAGE=0 -Wno-char-subscripts -Wno-parentheses-equality -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable     -Wno-undefined-bool-conversion -O2 -DTOP_TYPE=VORChain -include VORChain.h   -c -o VORChain__ALLcls.o VORChain__ALLcls.cpp
g++  -I.  -MMD -I/usr/local/share/verilator/include -I/usr/local/share/verilator/include/vltstd -DVL_PRINTF=printf -DVM_TRACE=1 -DVM_COVERAGE=0 -Wno-char-subscripts -Wno-parentheses-equality -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable     -Wno-undefined-bool-conversion -O2 -DTOP_TYPE=VORChain -include VORChain.h   -c -o VORChain__ALLsup.o VORChain__ALLsup.cpp
      Archiving VORChain__ALL.a ...
ar r VORChain__ALL.a VORChain__ALLcls.o VORChain__ALLsup.o
ar: creating VORChain__ALL.a
ranlib VORChain__ALL.a
g++    ORChain-harness.o verilated.o verilated_vcd_c.o VORChain__ALL.a    -o VORChain -lm -lstdc++  2>&1 | c++filt
make: Leaving directory `/home/smburns/bug-parameterized-dut/test_run_dir/bug.ORChain'
sim start on smburns-VirtualBox at Fri Sep 16 15:03:52 2016
inChannelName: 00009370.in
outChannelName: 00009370.out
cmdChannelName: 00009370.cmd
STARTING test_run_dir/bug.ORChain/VORChain
SEED 1474063432674
STEP 0 -> 1
  POKE ORChain.io_a_1 <- 0x0
  POKE ORChain.io_a_0 <- 0x0
  EXPECT ORChain.io_co_0 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_1 -> 0x0 == 0x0 PASS
STEP 1 -> 2
  POKE ORChain.io_a_1 <- 0x0
  POKE ORChain.io_a_0 <- 0x1
  EXPECT ORChain.io_co_0 -> 0x1 == 0x1 PASS
  EXPECT ORChain.io_co_1 -> 0x1 == 0x1 PASS
STEP 2 -> 3
  POKE ORChain.io_a_1 <- 0x1
  POKE ORChain.io_a_0 <- 0x0
  EXPECT ORChain.io_co_0 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_1 -> 0x1 == 0x1 PASS
STEP 3 -> 4
  POKE ORChain.io_a_1 <- 0x1
  POKE ORChain.io_a_0 <- 0x1
  EXPECT ORChain.io_co_0 -> 0x1 == 0x1 PASS
  EXPECT ORChain.io_co_1 -> 0x1 == 0x1 PASS
Enabling waves..
RAN 4 CYCLES PASSED
[info] ORChainTest2:
[info] ORChain of size 2
[info] - should work with various inputs
[info]   + Given a ORChain module 
[info]   + When InputArray a is Array(0, 0) 
[info]   + And After 1 cycle 
[info]   + Then OutputArray co is Array(0, 0) 
[info]   + When InputArray a is Array(1, 0) 
[info]   + And After 1 cycle 
[info]   + Then OutputArray co is Array(1, 1) 
[info]   + When InputArray a is Array(0, 1) 
[info]   + And After 1 cycle 
[info]   + Then OutputArray co is Array(0, 1) 
[info]   + When InputArray a is Array(1, 1) 
[info]   + And After 1 cycle 
[info]   + Then OutputArray co is Array(1, 1) 
[info] [0.000] Elaborating design...
[info] [0.017] Done elaborating.
verilator --cc ORChain.v --assert -Wno-fatal -Wno-WIDTH -Wno-STMTDLY --trace -O2 --top-module ORChain +define+TOP_TYPE=VORChain +define+PRINTF_COND=!ORChain.reset +define+STOP_COND=!ORChain.reset -CFLAGS -Wno-undefined-bool-conversion -O2 -DTOP_TYPE=VORChain -include VORChain.h -Mdir test_run_dir/bug.ORChain --exe ORChain-harness.cpp
make: Entering directory `/home/smburns/bug-parameterized-dut/test_run_dir/bug.ORChain'
g++  -I.  -MMD -I/usr/local/share/verilator/include -I/usr/local/share/verilator/include/vltstd -DVL_PRINTF=printf -DVM_TRACE=1 -DVM_COVERAGE=0 -Wno-char-subscripts -Wno-parentheses-equality -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable     -Wno-undefined-bool-conversion -O2 -DTOP_TYPE=VORChain -include VORChain.h   -c -o ORChain-harness.o ORChain-harness.cpp
ORChain-harness.cpp: In member function ‘void ORChain_api_t::init_sim_data()’:
ORChain-harness.cpp:24:61: error: ‘class VORChain’ has no member named ‘io_a_2’
         sim_data.inputs.push_back(new VerilatorCData(&(dut->io_a_2)));
                                                             ^
ORChain-harness.cpp:25:61: error: ‘class VORChain’ has no member named ‘io_a_3’
         sim_data.inputs.push_back(new VerilatorCData(&(dut->io_a_3)));
                                                             ^
ORChain-harness.cpp:28:62: error: ‘class VORChain’ has no member named ‘io_co_2’
         sim_data.outputs.push_back(new VerilatorCData(&(dut->io_co_2)));
                                                              ^
ORChain-harness.cpp:29:62: error: ‘class VORChain’ has no member named ‘io_co_3’
         sim_data.outputs.push_back(new VerilatorCData(&(dut->io_co_3)));
                                                              ^
At global scope:
cc1plus: warning: unrecognized command line option "-Wno-undefined-bool-conversion" [enabled by default]
cc1plus: warning: unrecognized command line option "-Wno-parentheses-equality" [enabled by default]
make: *** [ORChain-harness.o] Error 1
make: Leaving directory `/home/smburns/bug-parameterized-dut/test_run_dir/bug.ORChain'
sim start on smburns-VirtualBox at Fri Sep 16 15:03:54 2016
inChannelName: 00009388.in
outChannelName: 00009388.out
cmdChannelName: 00009388.cmd
STARTING test_run_dir/bug.ORChain/VORChain
SEED 1474063434297
STEP 0 -> 1
  POKE ORChain.io_a_1 <- 0x0
  POKE ORChain.io_a_0 <- 0x0
  POKE ORChain.io_a_3 <- 0x0
  POKE ORChain.io_a_2 <- 0x0
  EXPECT ORChain.io_co_0 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_1 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_2 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_3 -> 0x0 == 0x0 PASS
STEP 1 -> 2
  POKE ORChain.io_a_1 <- 0x0
  POKE ORChain.io_a_0 <- 0x1
  POKE ORChain.io_a_3 <- 0x0
  POKE ORChain.io_a_2 <- 0x0
  EXPECT ORChain.io_co_0 -> 0x1 == 0x1 PASS
  EXPECT ORChain.io_co_1 -> 0x1 == 0x1 PASS
  EXPECT ORChain.io_co_2 -> 0x0 == 0x1 FAIL
  EXPECT ORChain.io_co_3 -> 0x0 == 0x1 FAIL
STEP 2 -> 3
  POKE ORChain.io_a_1 <- 0x1
  POKE ORChain.io_a_0 <- 0x0
  POKE ORChain.io_a_3 <- 0x0
  POKE ORChain.io_a_2 <- 0x0
  EXPECT ORChain.io_co_0 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_1 -> 0x1 == 0x1 PASS
  EXPECT ORChain.io_co_2 -> 0x0 == 0x1 FAIL
  EXPECT ORChain.io_co_3 -> 0x0 == 0x1 FAIL
STEP 3 -> 4
  POKE ORChain.io_a_1 <- 0x0
  POKE ORChain.io_a_0 <- 0x0
  POKE ORChain.io_a_3 <- 0x0
  POKE ORChain.io_a_2 <- 0x1
  EXPECT ORChain.io_co_0 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_1 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_2 -> 0x0 == 0x1 FAIL
  EXPECT ORChain.io_co_3 -> 0x0 == 0x1 FAIL
STEP 4 -> 5
  POKE ORChain.io_a_1 <- 0x0
  POKE ORChain.io_a_0 <- 0x0
  POKE ORChain.io_a_3 <- 0x1
  POKE ORChain.io_a_2 <- 0x0
  EXPECT ORChain.io_co_0 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_1 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_2 -> 0x0 == 0x0 PASS
  EXPECT ORChain.io_co_3 -> 0x0 == 0x1 FAIL
Enabling waves..
RAN 5 CYCLES FAILED FIRST AT CYCLE 2
[info] ORChainTest4:
[info] ORChain of size 4
[info] - should work with various inputs *** FAILED ***
[info]   false was not true (bug.scala:116)
[info]   + Given a ORChain module 
[info]   + When InputArray a is Array(0, 0, 0, 0) 
[info]   + And After 1 cycle 
[info]   + Then OutputArray co is Array(0, 0, 0, 0) 
[info]   + When InputArray a is Array(1, 0, 0, 0) 
[info]   + And After 1 cycle 
[info]   + Then OutputArray co is Array(1, 1, 1, 1) 
[info]   + When InputArray a is Array(0, 1, 0, 0) 
[info]   + And After 1 cycle 
[info]   + Then OutputArray co is Array(0, 1, 1, 1) 
[info]   + When InputArray a is Array(0, 0, 1, 0) 
[info]   + And After 1 cycle 
[info]   + Then OutputArray co is Array(0, 0, 1, 1) 
[info]   + When InputArray a is Array(0, 0, 0, 1) 
[info]   + And After 1 cycle 
[info]   + Then OutputArray co is Array(0, 0, 0, 1) 
[info] ScalaCheck
[info] Passed: Total 0, Failed 0, Errors 0, Passed 0
[info] ScalaTest
[info] Run completed in 11 seconds, 89 milliseconds.
[info] Total number of tests run: 2
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 1, failed 1, canceled 0, ignored 0, pending 0
[info] *** 1 TEST FAILED ***
[error] Failed: Total 2, Failed 1, Errors 0, Passed 1
[error] Failed tests:
[error] 	bug.ORChainTest4
[error] (test:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 14 s, completed Sep 16, 2016 3:03:55 PM
````

--- a/src/main/scala/chisel3/iotesters/Backends.scala
+++ b/src/main/scala/chisel3/iotesters/Backends.scala
@@ -8,7 +8,7 @@ import java.io.PrintStream
   * define interface for ClassicTester backend implementations such as verilator and firrtl interpreter
   */
 
-private[iotesters] abstract class Backend(_seed: Long = System.currentTimeMillis) {
+private[iotesters] abstract class Backend(private[iotesters] val _seed: Long = System.currentTimeMillis) {
   val rnd = new scala.util.Random(_seed)
 
   def poke(signal: InstanceId, value: BigInt, off: Option[Int])

--- a/src/main/scala/chisel3/iotesters/ChiselMain.scala
+++ b/src/main/scala/chisel3/iotesters/ChiselMain.scala
@@ -16,7 +16,8 @@ private[iotesters] class TesterContext {
   var isRunTest = false
   var testerSeed = System.currentTimeMillis
   val testCmd = ArrayBuffer[String]()
-  var backend = "verilator"
+  var backendType = "verilator"
+  var backend: Option[Backend] = None
   var targetDir = new File("test_run_dir")
   var logFile: Option[File] = None
   var waveform: Option[File] = None
@@ -27,12 +28,12 @@ object chiselMain {
   private[iotesters] def context = contextVar.value getOrElse (new TesterContext)
 
   private def parseArgs(args: List[String]): Unit = args match {
-    case "--firrtl" :: tail => context.backend = "firrtl" ; parseArgs(tail)
-    case "--verilator" :: tail => context.backend = "verilator" ; parseArgs(tail)
-    case "--vcs" :: tail => context.backend = "vcs" ; parseArgs(tail)
-    case "--glsim" :: tail => context.backend = "glsim" ; parseArgs(tail)
+    case "--firrtl" :: tail => context.backendType = "firrtl" ; parseArgs(tail)
+    case "--verilator" :: tail => context.backendType = "verilator" ; parseArgs(tail)
+    case "--vcs" :: tail => context.backendType = "vcs" ; parseArgs(tail)
+    case "--glsim" :: tail => context.backendType = "glsim" ; parseArgs(tail)
     case "--v" :: tail  => context.isGenVerilog = true ; parseArgs(tail)
-    case "--backend" :: value :: tail => context.backend = value ; parseArgs(tail)
+    case "--backend" :: value :: tail => context.backendType = value ; parseArgs(tail)
     case "--genHarness" :: tail => context.isGenHarness = true ; parseArgs(tail)
     case "--compile" :: tail => context.isCompiling = true ; parseArgs(tail)
     case "--test" :: tail => context.isRunTest = true ; parseArgs(tail)
@@ -47,7 +48,7 @@ object chiselMain {
 
   private def genHarness[T <: Module](dut: Module, nodes: Seq[internal.InstanceId], chirrtl: firrtl.ir.Circuit) {
     val dir = context.targetDir
-    context.backend match {
+    context.backendType match {
       case "firrtl" => // skip
       case "verilator" =>
         val harness = new FileWriter(new File(dir, s"${chirrtl.main}-harness.cpp"))
@@ -58,14 +59,14 @@ object chiselMain {
       case "vcs" | "glsim" =>
         val harness = new FileWriter(new File(dir, s"${chirrtl.main}-harness.v"))
         val waveform = (new File(dir, s"${chirrtl.main}.vpd")).toString
-        genVCSVerilogHarness(dut, harness, waveform.toString, context.backend == "glsim")
+        genVCSVerilogHarness(dut, harness, waveform.toString, context.backendType == "glsim")
       case b => throw BackendException(b)
     }
   }
 
   private def compile(dutName: String) {
     val dir = context.targetDir
-    context.backend match {
+    context.backendType match {
       case "firrtl" => // skip
       case "verilator" =>
         // Copy API files
@@ -101,24 +102,32 @@ object chiselMain {
     val chirrtl = firrtl.Parser.parse(chisel3.Driver.emit(circuit))
     val chirrtlFile = new File(dir, s"${name}.ir")
     val verilogFile = new File(dir, s"${name}.v")
-    if (context.backend == "firrtl") {
-       val writer = new FileWriter(chirrtlFile)
-      firrtl.FIRRTLEmitter run (chirrtl, writer)
-      writer.close
-    } else if (context.isGenVerilog) {
-      val annotation = new firrtl.Annotations.AnnotationMap(Seq(
-        new firrtl.passes.InferReadWriteAnnotation(name, firrtl.Annotations.TransID(-1))))
-      val writer = new FileWriter(verilogFile)
-      new firrtl.VerilogCompiler compile (chirrtl, annotation, writer)
-      writer.close
+    context.backendType match {
+      case "firrtl" =>
+        val writer = new FileWriter(chirrtlFile)
+        firrtl.FIRRTLEmitter run (chirrtl, writer)
+        writer.close
+      case _ if (context.isGenVerilog) =>
+        val annotation = new firrtl.Annotations.AnnotationMap(Seq(
+          new firrtl.passes.InferReadWriteAnnotation(name, firrtl.Annotations.TransID(-1))))
+        val writer = new FileWriter(verilogFile)
+        new firrtl.VerilogCompiler compile (chirrtl, annotation, writer)
+        writer.close
+      case _ =>
     } 
 
     if (context.isGenHarness) genHarness(dut, nodes, chirrtl)
 
     if (context.isCompiling) compile(name)
 
+    dut
+  }
+
+  private def setupBackend[T <: Module](dut: T) {
+    val name = dut.name
+
     if (context.testCmd.isEmpty) {
-      context.backend match {
+      context.backendType match {
         case "firrtl" => // skip
         case "verilator" =>
           context.testCmd += (new File(context.targetDir, s"V$name")).toString
@@ -127,11 +136,23 @@ object chiselMain {
         case b => throw BackendException(b)
       }
     }
+
     context.waveform match {
       case None =>
       case Some(f) => context.testCmd += s"+waveform=$f"
     }
-    dut
+
+    context.backend = Some(context.backendType match {
+      case "firrtl" =>
+        val file = new java.io.File(context.targetDir, s"${dut.name}.ir")
+        val ir = io.Source.fromFile(file).getLines mkString "\n"
+        new FirrtlTerpBackend(dut, ir, context.testerSeed)
+      case "verilator" =>
+        new VerilatorBackend(dut, context.testCmd.toList, context.testerSeed)
+      case "vcs" | "glsim" =>
+        new VCSBackend(dut, context.testCmd.toList, context.testerSeed)
+      case b => throw BackendException(b)
+    })
   }
 
   def apply[T <: Module](args: Array[String], dutGen: () => T): T = {
@@ -147,11 +168,18 @@ object chiselMain {
     contextVar.withValue(Some(new TesterContext)) {
       val dut = elaborate(args, dutGen)
       if (context.isRunTest) {
+        setupBackend(dut)
         assert(try {
           testerGen(dut).finish
         } catch { case e: Throwable =>
           e.printStackTrace
-          TesterProcess.killall
+          context.backend match {
+            case Some(b: VCSBackend) =>
+              TesterProcess kill b
+            case Some(b: VerilatorBackend) =>
+              TesterProcess kill b
+            case _ =>
+          }
           false
         }, "Test failed")
       }

--- a/src/main/scala/chisel3/iotesters/Driver.scala
+++ b/src/main/scala/chisel3/iotesters/Driver.scala
@@ -20,18 +20,24 @@ object Driver {
                          backendType: String = "firrtl",
                          dir: File = createTempDirectory("test-out"))
                          (testerGen: T => PeekPokeTester[T]): Boolean = {
-    val (dut, backend) = backendType match {
+    val (dut, b) = backendType match {
       case "firrtl" => setupFirrtlTerpBackend(dutGen, dir)
       case "verilator" => setupVerilatorBackend(dutGen, dir)
       case "vcs" => setupVCSBackend(dutGen, dir)
       case _ => throw new Exception("Unrecongnized backend type $backendType")
     }
-    backendVar.withValue(Some(backend)) {
+    backendVar.withValue(Some(b)) {
       try {
         testerGen(dut).finish
       } catch { case e: Throwable =>
         e.printStackTrace
-        TesterProcess.killall
+        backend match {
+          case Some(b: VCSBackend) =>
+            TesterProcess kill b
+          case Some(b: VerilatorBackend) =>
+            TesterProcess kill b
+          case _ =>
+        }
         throw e
       }
     }
@@ -50,7 +56,13 @@ object Driver {
         testerGen(dut).finish
       } catch { case e: Throwable =>
         e.printStackTrace
-        TesterProcess.killall
+        backend match {
+          case Some(b: VCSBackend) =>
+            TesterProcess kill b
+          case Some(b: VerilatorBackend) =>
+            TesterProcess kill b
+          case _ =>
+        }
         throw e
       }
     }

--- a/src/main/scala/chisel3/iotesters/Driver.scala
+++ b/src/main/scala/chisel3/iotesters/Driver.scala
@@ -3,6 +3,7 @@
 package chisel3.iotesters
 
 import chisel3.Module
+import chisel3.Driver.createTempDirectory
 import scala.util.DynamicVariable
 import java.io.File
 
@@ -15,12 +16,14 @@ object Driver {
     * @@backendType determines whether the ClassicTester uses verilator or the firrtl interpreter to simulate the circuit
     * Will do intermediate compliation steps to setup the backend specified, including cpp compilation for the verilator backend and firrtl IR compilation for the firrlt backend
     */
-  def apply[T <: Module](dutGen: () => T, backendType: String = "firrtl")(
-      testerGen: T => PeekPokeTester[T]): Boolean = {
+  def apply[T <: Module](dutGen: () => T,
+                         backendType: String = "firrtl",
+                         dir: File = createTempDirectory("test-out"))
+                         (testerGen: T => PeekPokeTester[T]): Boolean = {
     val (dut, backend) = backendType match {
-      case "firrtl" => setupFirrtlTerpBackend(dutGen)
-      case "verilator" => setupVerilatorBackend(dutGen)
-      case "vcs" => setupVCSBackend(dutGen)
+      case "firrtl" => setupFirrtlTerpBackend(dutGen, dir)
+      case "verilator" => setupVerilatorBackend(dutGen, dir)
+      case "vcs" => setupVCSBackend(dutGen, dir)
       case _ => throw new Exception("Unrecongnized backend type $backendType")
     }
     backendVar.withValue(Some(backend)) {

--- a/src/main/scala/chisel3/iotesters/FirrtlTerpBackend.scala
+++ b/src/main/scala/chisel3/iotesters/FirrtlTerpBackend.scala
@@ -87,10 +87,9 @@ private[iotesters] class FirrtlTerpBackend(dut: Module,
 }
 
 private[iotesters] object setupFirrtlTerpBackend {
-  def apply[T <: chisel3.Module](dutGen: () => T): (T, Backend) = {
+  def apply[T <: chisel3.Module](dutGen: () => T, dir: File): (T, Backend) = {
     val circuit = chisel3.Driver.elaborate(dutGen)
     val dut = getTopModule(circuit).asInstanceOf[T]
-    val dir = new File(s"test_run_dir/${dut.getClass.getName}") ; dir.mkdirs()
 
     // Dump FIRRTL for debugging
     chisel3.Driver.dumpFirrtl(circuit, Some(new File(dir, s"${circuit.name}.fir")))

--- a/src/main/scala/chisel3/iotesters/PeekPokeTester.scala
+++ b/src/main/scala/chisel3/iotesters/PeekPokeTester.scala
@@ -29,8 +29,7 @@ trait PeekPokeTests {
 abstract class PeekPokeTester[+T <: Module](val dut: T,
                                             verbose: Boolean = true,
                                             base: Int = 16,
-                                            logFile: Option[java.io.File] = None,
-                                            _seed: Long = chiselMain.context.testerSeed) {
+                                            logFile: Option[java.io.File] = None) {
 
   implicit def longToInt(x: Long) = x.toInt
 
@@ -49,21 +48,11 @@ abstract class PeekPokeTester[+T <: Module](val dut: T,
   /****************************/
   /*** Simulation Interface ***/
   /****************************/
-  logger println s"SEED ${_seed}"
-  val backend = Driver.backend getOrElse {
-    val cmd = chiselMain.context.testCmd.toList
-    chiselMain.context.backend match {
-      case "firrtl" =>
-        val file = new java.io.File(chiselMain.context.targetDir, s"${dut.name}.ir")
-        val ir = io.Source.fromFile(file).getLines mkString "\n"
-        new FirrtlTerpBackend(dut, ir, _seed)
-      case "verilator" =>
-        new VerilatorBackend(dut, cmd, _seed)
-      case "vcs" | "glsim" =>
-        new VCSBackend(dut, cmd, _seed)
-      case b => throw BackendException(b)
-    }
+  val backend = ((Driver.backend, chiselMain.context.backend): @unchecked) match {
+    case (Some(b), _) => b
+    case (None, Some(b)) => b
   }
+  logger println s"SEED ${backend._seed}"
 
   /********************************/
   /*** Classic Tester Interface ***/

--- a/src/main/scala/chisel3/iotesters/PeekPokeTesterUtils.scala
+++ b/src/main/scala/chisel3/iotesters/PeekPokeTesterUtils.scala
@@ -106,32 +106,20 @@ private[iotesters] case class TestApplicationException(exitVal: Int, lastMessage
   extends RuntimeException(lastMessage)
 
 private[iotesters] object TesterProcess {
-  val processes = ArrayBuffer[Process]()
-
-  def apply(cmd: Seq[String], logs: ArrayBuffer[String]) = synchronized {
+  def apply(cmd: Seq[String], logs: ArrayBuffer[String]) = {
     require(new java.io.File(cmd.head).exists, s"${cmd.head} doesn't exists")
     val processBuilder = Process(cmd mkString " ")
     val processLogger = ProcessLogger(println, logs += _) // don't log stdout
-    val process = processBuilder run processLogger
-    processes += process
-    process
+    processBuilder run processLogger
   }
-
-  def finish(p: Process) = synchronized {
-    processes -= p
-    // p.destroy
+  def kill(sim: SimApiInterface) {
+    while(!sim.exitValue.isCompleted) sim.process.destroy
+    println("Exit Code: %d".format(sim.process.exitValue))
   }
-
-  import scala.concurrent.{Future, Await, blocking}
-  import scala.concurrent.duration._
-  import scala.concurrent.ExecutionContext.Implicits.global
-  def killall = synchronized {
-    processes map { p =>
-      val exitValue = Future(blocking(p.exitValue))
-      while(!exitValue.isCompleted) p.destroy
-      println("Exit Code: %d".format(
-        Await.result(exitValue, Duration(-1, SECONDS))))
-    }
-    processes.clear
+  def kill(p: VCSBackend) {
+    kill(p.simApiInterface)
+  }
+  def kill(p: VerilatorBackend) {
+    kill(p.simApiInterface)
   }
 }

--- a/src/main/scala/chisel3/iotesters/SimApiInterface.scala
+++ b/src/main/scala/chisel3/iotesters/SimApiInterface.scala
@@ -6,8 +6,7 @@ import chisel3._
 
 import scala.collection.mutable.{ArrayBuffer, HashMap}
 import scala.collection.immutable.ListMap
-import scala.concurrent.ExecutionContext.Implicits.global
-import scala.concurrent.{Future, Await, blocking}
+import scala.concurrent.{Future, Await, ExecutionContext, blocking}
 import scala.concurrent.duration._
 import scala.sys.process.{Process, ProcessLogger}
 import java.io.{File, PrintStream}
@@ -32,6 +31,47 @@ private[iotesters] class SimApiInterface(dut: Module, cmd: Seq[String]) {
   private val _signalMap = HashMap[String, Int]()
   private val _chunks = HashMap[String, Int]()
   private val _logs = ArrayBuffer[String]()
+
+  //initialize simulator process
+  private[iotesters] val process = TesterProcess(cmd, _logs)
+  // Set up a Future to wait for (and signal) the test process exit.
+  import ExecutionContext.Implicits.global
+  private[iotesters] val exitValue = Future(blocking(process.exitValue))
+  // memory mapped channels
+  private val (inChannel, outChannel, cmdChannel) = {
+    // Wait for the startup message
+    // NOTE: There may be several messages before we see our startup message.
+    val simStartupMessageStart = "sim start on "
+    while (!_logs.exists(_ startsWith simStartupMessageStart) && !exitValue.isCompleted) { Thread.sleep(100) }
+    // Remove the startup message (and any precursors).
+    while (!_logs.isEmpty && !_logs.head.startsWith(simStartupMessageStart)) {
+      println(_logs.remove(0))
+    }
+    println(if (!_logs.isEmpty) _logs.remove(0) else "<no startup message>")
+    while (_logs.size < 3) {
+      // If the test application died, throw a run-time error.
+      throwExceptionIfDead(exitValue)
+      Thread.sleep(100)
+    }
+    val in_channel_name = _logs.remove(0)
+    val out_channel_name = _logs.remove(0)
+    val cmd_channel_name = _logs.remove(0)
+    val in_channel = new Channel(in_channel_name)
+    val out_channel = new Channel(out_channel_name)
+    val cmd_channel = new Channel(cmd_channel_name)
+
+    println(s"inChannelName: ${in_channel_name}")
+    println(s"outChannelName: ${out_channel_name}")
+    println(s"cmdChannelName: ${cmd_channel_name}")
+
+    in_channel.consume
+    cmd_channel.consume
+    in_channel.release
+    out_channel.release
+    cmd_channel.release
+
+    (in_channel, out_channel, cmd_channel)
+  }
 
   private def dumpLogs(implicit logger: PrintStream) {
     _logs foreach logger.println
@@ -266,56 +306,12 @@ private[iotesters] class SimApiInterface(dut: Module, cmd: Seq[String]) {
 
   def finish(implicit logger: PrintStream) {
     mwhile(!sendCmd(SIM_CMD.FIN)) { }
-    while(!exitValue.isCompleted) { }
+    println("Exit Code: %d".format(
+      Await.result(exitValue, Duration.Inf)))
     dumpLogs
     inChannel.close
     outChannel.close
     cmdChannel.close
-    TesterProcess.finish(process)
-  }
-
-  //initialize cpp process
-  private val process = TesterProcess(cmd, _logs)
-  // Set up a Future to wait for (and signal) the test process exit.
-  private val exitValue: Future[Int] = Future {
-    blocking {
-      process.exitValue
-    }
-  }
-  // memory mapped channels
-  private val (inChannel, outChannel, cmdChannel) = {
-    // Wait for the startup message
-    // NOTE: There may be several messages before we see our startup message.
-    val simStartupMessageStart = "sim start on "
-    while (!_logs.exists(_ startsWith simStartupMessageStart) && !exitValue.isCompleted) { Thread.sleep(100) }
-    // Remove the startup message (and any precursors).
-    while (!_logs.isEmpty && !_logs.head.startsWith(simStartupMessageStart)) {
-      println(_logs.remove(0))
-    }
-    println(if (!_logs.isEmpty) _logs.remove(0) else "<no startup message>")
-    while (_logs.size < 3) {
-      // If the test application died, throw a run-time error.
-      throwExceptionIfDead(exitValue)
-      Thread.sleep(100)
-    }
-    val in_channel_name = _logs.remove(0)
-    val out_channel_name = _logs.remove(0)
-    val cmd_channel_name = _logs.remove(0)
-    val in_channel = new Channel(in_channel_name)
-    val out_channel = new Channel(out_channel_name)
-    val cmd_channel = new Channel(cmd_channel_name)
-
-    println(s"inChannelName: ${in_channel_name}")
-    println(s"outChannelName: ${out_channel_name}")
-    println(s"cmdChannelName: ${cmd_channel_name}")
-
-    in_channel.consume
-    cmd_channel.consume
-    in_channel.release
-    out_channel.release
-    cmd_channel.release
-
-    (in_channel, out_channel, cmd_channel)
   }
 
   // Once everything has been prepared, we can start the communications.

--- a/src/main/scala/chisel3/iotesters/VCSBackend.scala
+++ b/src/main/scala/chisel3/iotesters/VCSBackend.scala
@@ -113,10 +113,9 @@ object genVCSVerilogHarness {
 }
 
 private[iotesters] object setupVCSBackend {
-  def apply[T <: chisel3.Module](dutGen: () => T): (T, Backend) = {
+  def apply[T <: chisel3.Module](dutGen: () => T, dir: File): (T, Backend) = {
     val circuit = chisel3.Driver.elaborate(dutGen)
     val dut = getTopModule(circuit).asInstanceOf[T]
-    val dir = new File(s"test_run_dir/${dut.getClass.getName}") ; dir.mkdirs()
 
     // Generate CHIRRTL
     val chirrtl = firrtl.Parser.parse(chisel3.Driver.emit(dutGen))

--- a/src/main/scala/chisel3/iotesters/VerilatorBackend.scala
+++ b/src/main/scala/chisel3/iotesters/VerilatorBackend.scala
@@ -265,13 +265,12 @@ class VerilatorCppHarnessCompiler(dut: Chisel.Module,
 }
 
 private[iotesters] object setupVerilatorBackend {
-  def apply[T <: chisel3.Module](dutGen: () => T): (T, Backend) = {
+  def apply[T <: chisel3.Module](dutGen: () => T, dir: File): (T, Backend) = {
     // Generate CHIRRTL
     val circuit = chisel3.Driver.elaborate(dutGen)
     val chirrtl = firrtl.Parser.parse(chisel3.Driver.emit(circuit))
     val dut = getTopModule(circuit).asInstanceOf[T]
     val nodes = getChiselNodes(circuit)
-    val dir = new File(s"test_run_dir/${dut.getClass.getName}"); dir.mkdirs()
 
     // Generate Verilog
     val verilogFile = new File(dir, s"${circuit.name}.v")

--- a/src/main/scala/chisel3/iotesters/VerilatorBackend.scala
+++ b/src/main/scala/chisel3/iotesters/VerilatorBackend.scala
@@ -299,7 +299,7 @@ private[iotesters] class VerilatorBackend(dut: Chisel.Module,
                                           cmd: Seq[String],
                                           _seed: Long = System.currentTimeMillis) extends Backend(_seed) {
 
-  val simApiInterface = new SimApiInterface(dut, cmd)
+  private[iotesters] val simApiInterface = new SimApiInterface(dut, cmd)
 
   def poke(signal: InstanceId, value: BigInt, off: Option[Int])
           (implicit logger: PrintStream, verbose: Boolean, base: Int) {

